### ISSUE TITLE
Parse error: on input `='

### ISSUE THREAD

**Issue by asarkar** [ndmitchell/hlint#1402]
The following function is part of a module.
```
parse :: String -> Trio
parse text = (n, op, s)
  where
    pattern = "[-]?[[:digit:]]+"
    (prefix, num, suffix) = text =~ pattern :: (String, String, String)
    trim = T.unpack . T.strip . T.pack
    n = R.readMaybe $ trim num
    op = trim prefix
    s = trim suffix
```

This compiles and works for the given test cases, but hlint emits an error.
```
Error: Parse error: on input `='
Found:
    parse text = (n, op, s)
      where
  >     pattern = "[-]?[[:digit:]]+"
        (prefix, num, suffix) = text =~ pattern :: (String, String, String)
        trim = T.unpack . T.strip . T.pack 
 ```
 
I've gone through most of the [tickets that reported similar errors](https://github.com/ndmitchell/hlint/issues?q=parse+error+on+input), and the suggestion is to disable `TemplateHaskell` using the pragma `{-# LANGUAGE NoTemplateHaskell #-}`. However, that doesn't help in this case.

**Comment by asarkar** [comment]
I can confirm that using the pragma `{-# LANGUAGE NoPatternSynonyms #-}` fixes the error. Can hlint not turn on language extensions magically?

**Comment by ndmitchell** [comment]
@asarkar HLint could do anything it wanted. But it's a balance - do we break more users by turning on PatternSynonyms, or fix more users? I think the view is that it fixes more users than it breaks, so on by default is the right default. If you pass `-XHaskell2020` or similar I think it restricts you to a specific language standard, so you can opt in to following a standard rather than the default behaviour.

**Comment by asarkar** [comment]
What I don’t understand, perhaps from my lack of experience working with Haskell, is that why hlint can’t use the user-provided configuration, if any, instead of imposing its own opinion about what’s good for the user?

**Comment by shayne-fletcher** [comment]
hlint will use the user provided configuration if it is given one. as @ndmitchell  recommend pass `-XGhc2021` (or `-XHaskell2010 or `-XHaskell98`) to hlint. doing this you have precisely specified what base set of extensions are enabled.

**Comment by asarkar** [comment]
> GHC2021 is used by GHC if neither Haskell98 nor Haskell2010 is turned on explicitly.

It seems to be the default without me having to specify explicitly. That’s the way I’d like to keep it, so that the code is always compatible with the latest language specification.

**Comment by shayne-fletcher** [comment]
> > GHC2021 is used by GHC if neither Haskell98 nor Haskell2010 is turned on explicitly.
> 
> It seems to be the default without me having to specify explicitly. That’s the way I’d like to keep it, so that the code is always compatible with the latest language specification. Notably, `PatternSynonyms` is _not_ enabled by GHC2021.

if no language specifier is provide hlint constructs its own set of extensions to default enable. that set enables more extensions than GHC would default enable if not invoked with an explicit language specifier.

**Comment by shayne-fletcher** [comment]
the default behavior is chosen for hlint is to parse as many files as possible (i.e. default behavior = not give a language specifier).

**Comment by asarkar** [comment]
> if no language specifier is provided hlint constructs its own set of extensions

... which is different from what GHC chooses, hence the problem. Why not choose the default the same as GHC?

> hlint to parse as many files as possible

Is there any substantiated data that shows the difference between hlint chosen extensions and GHC chosen extensions? In other words, can it be shown that the if hlint chose the same default extensions as GHC, it'd fail to parse x% of files?

**Comment by shayne-fletcher** [comment]
> > if no language specifier is provided hlint constructs its own set of extensions
> 
> ... which is different from what GHC chooses, hence the problem. Why not choose the default the same as GHC?

i agree this a reasonable proposition.

> > hlint to parse as many files as possible
> 
> Is there any substantiated data that shows the difference between hlint chosen extensions and GHC chosen extensions? In other words, can it be shown that the if hlint chose the same default extensions as GHC, it'd fail to parse x% of files?

i don't know about that.

**Comment by asarkar** [comment]
Now that we have established this ticket has a valid point, I’ve updated the title to better suit the problem.

**Comment by asarkar** [comment]
> turn on most

There’s a definitive list of the extensions enabled by default by GHC. What this ticket is asking is for hlint to use that list instead of making up its own. Is that what you referred to as hard, almost impossible?

**Comment by ndmitchell** [comment]
No, the list of extensions enabled by GHC _in your project_ is based on those defined in Cabal, Stack, command line flags and elsewhere. We can get the list of extensions enabled if you type `ghc`, but since that doesn't work in most projects, it's not that helpful. What we want is to approximate the extensions enabled in your project.

**Comment by ndmitchell** [comment]
I'm not sure what you mean. Consider the case that Steve has a project. In his `.cabal` file he has the extension `PatternSynonyms` enabled. Currently HLint works. If I follow the same default as GHC, it would not. Audrey has a project, he uses a custom build system like Bazel, and in herproject he has `TemplateHaskell` turned on in the `.bazelconfig` that applies to every project in her megarepo. Currently HLint works, but with the same default as GHC, it doesn't. My understanding based on talking to users is that if HLint defaulted to what GHC does (which you can do with a `.hlint.yaml` file if you want), then fewer projects would work out of the box.

**Comment by asarkar** [comment]
>  In his .cabal file he has the extension PatternSynonyms enabled. Currently HLint works. If I follow the same default as GHC, it would not

If user has enabled any extensions, then hlint won't fallback to defaults. All I'm saying is that when the user hasn't enabled any extensions, GHC defaults and hlint defaults aren't the same.

**Comment by ndmitchell** [comment]
I don't follow. In the .cabal file the user has enabled PatternSynonyms. What should HLint do?

> If user has enabled any extensions, then hlint won't fallback to defaults

How do we detect if a user has enabled any extensions in their .cabal/stack/bazel projects?
